plugins/image-prioritizer/load.php

```diff
@@ -44,9 +44,14 @@ static function ( string $global_var_name, string $version, Closure $load ): voi
 				}
 			};
 
-			// Wait until after the plugins have loaded and the theme has loaded. The after_setup_theme action is used
-			// because it is the first action that fires once the theme is loaded.
-			add_action( 'after_setup_theme', $bootstrap, PHP_INT_MIN );
+			/*
+			 * Wait until after the plugins have loaded and the theme has loaded. The after_setup_theme action could be
+			 * used since it is the first action that fires once the theme is loaded. However, plugins may embed this
+			 * logic inside a module which initializes even later at the init action. The earliest action that this
+			 * plugin has hooks for is the init action at the default priority of 10 (which includes the rest_api_init
+			 * action), so this is why it gets initialized at priority 9.
+			 */
+			add_action( 'init', $bootstrap, 9 );
```

Comment:
I'm not sure I understand this. Why is the timing of this (in Image Prioritizer and Embed Optimizer) relevant for other plugins? Are you thinking of extensions to _those_ plugins (rather than extensions to Optimization Detective)?

Another thought: Why can't this use the `od_init` action?

Reply:
Bootstrapping at `init` is all about the embeddability of these plugins inside of other plugins, which is why this whole bootstrap logic was first devised (https://github.com/WordPress/performance/issues/1081, https://github.com/WordPress/performance/pull/1159, https://github.com/WordPress/performance/pull/1185). 

For example, in https://github.com/elementor/elementor/pull/28844 I've created a new module for Elementor that embeds the Optimization Detective plugin and its extensions. Elementor initializes its modules at `init` priority 0. Since none of the functionality in Optimization Detective needs to be initialized before `init`, there's no need for the current use of `after_setup_theme` to initialize. So this postpones initialization to latest possible point which opens the door to be embedded in more places.

Reply:
This still feels like a rather shaky foundation to me. For example, why does it use priority `9`?

I think I'm still not following what this is meant to do: I think I understand now why certain code needs to run before this callback, but why does certain other code need to run after this callback? For instance, why couldn't we use `init` on `999999`, or even `wp_loaded` or a later hook?

Reply:
Priority 9 is one less than the default priority of 10, as noted in the comment. There are callbacks added at this default priority in the plugin. So priority 9 is the last.

If we used `init` with an even later priority, then we'd need to adjust the other `init` actions' priorities as well in this plugin, for example:

https://github.com/WordPress/performance/blob/94fc5e7215516d58dc307f973cb75ae1432ad042/plugins/optimization-detective/storage/class-od-url-metrics-post-type.php#L48

https://github.com/WordPress/performance/blob/94fc5e7215516d58dc307f973cb75ae1432ad042/plugins/embed-optimizer/hooks.php#L27

If we went with `wp_loaded`, then we couldn't use the `init` action in the plugins at all, which as I understand is too late for when post types should be registered.

Reply:
So basically this needs to be called as late as possible, but before any of Optimization Detective's own hooks would fire?

Reply:
Yes, exactly. As late as possible, and hopefully after any other possible embedding plugins' modules initialize (e.g. Elementor initializes at `init` priority `0`).

Reply:
Makes sense. Since we have full control over the code in Optimization Detective, we may want to consider moving the plugin's `init` callbacks to a very late priority, so that we can also push this back even further, while still being on `init`. That way we can cater for other plugins even better, since I wouldn't find it unreasonable for other plugins to load modules on `init` with the default priority either.

Given the callbacks from Optimization Detective do not _have_ to run after everything else, maybe we just use a very high number (rather than `PHP_INT_MAX` which I would only use if it technically has to be the last thing).

We can open a follow up PR for this though, if you agree?

Reply:
Sounds good!